src/libmongoc/src/mongoc/mongoc-client-side-encryption.c

```diff
@@ -2672,8 +2674,182 @@ _mongoc_cse_is_enabled (mongoc_client_t *client)
+{
+   struct cec_context *cec = ctx->userdata;
+   return mongoc_client_encryption_create_datakey (cec->enc,
+                                                   cec->kms_provider,
+                                                   cec->dk_opts,
+                                                   ctx->out_keyid,
+                                                   ctx->out_error);
+}
+
+mongoc_collection_t *
+mongoc_client_encryption_create_encrypted_collection (
+   mongoc_client_encryption_t *enc,
+   mongoc_database_t *database,
+   const char *const name,
+   const bson_t *in_options,
+   bson_t *new_options,
+   const char *const kms_provider,
+   const mongoc_client_encryption_datakey_opts_t *dk_opts,
+   bson_error_t *error)
+{
+   bson_t local_new_options = BSON_INITIALIZER;
+   if (!new_options) {
+      // We'll use our own storage for the new options
+      new_options = &local_new_options;
+   }
+
+   // Init the storage. Either inits the caller's copy, or our local version.
+   bson_init (new_options);
+
+   // Context for the creation of a new datakey
+   struct cec_context ctx = {
+      .enc = enc,
+      .dk_opts = dk_opts,
+      .kms_provider = kms_provider,
+   };
+
+   mongoc_collection_t *ret = NULL;
+   // Create the new options, filling out the 'keyId' automatically:
+   if (_mongoc_cec_fill_auto_datakeys (
+          new_options, in_options, _auto_datakey, &ctx, error)) {
+      // We've successfully filled out all null keyIds. Now create the
+      // collection with our new options:
+      ret =
+         mongoc_database_create_collection (database, name, new_options, error);
+   }
+   // Destroy the local options, which may or may not have been used. If unused,
+   // the new options are now owned by the caller.
+   bson_destroy (&local_new_options);
+   // The resulting collection, or NULL on error:
+   return ret;
+}
+
 #endif /* MONGOC_ENABLE_CLIENT_SIDE_ENCRYPTION */
 
+/// Generate one encryptedField element.
+static void
+_init_1_encryptedField (bson_t *out_field,
+                        const bson_t *in_field,
+                        auto_datakey_factory fac,
+                        void *fac_userdata,
+                        bson_error_t *error)
+{
+   bsonVisitEach (
+      *in_field,
+      // If it is not a "keyId":null element, just copy it to the output.
+      if (not(keyWithType ("keyId", null)),
+          then (appendTo (*out_field), continue)),
+      // Otherwise:
+      do({
+         // Set up factory context
+         bson_value_t new_key = {0};
+         struct auto_datakey_context ctx = {
+            .out_keyid = &new_key,
+            .out_error = error,
+            .userdata = fac_userdata,
+         };
+         // Call the callback to create the new key
+         if (!fac (&ctx)) {
+            bsonParseError = "Factory function indicated failure";
+         } else {
+            // Append to the field
+            BSON_APPEND_VALUE (out_field, "keyId", &new_key);
```

Comment:
```suggestion
            BSON_APPEND_VALUE (out_field, "keyId", &new_key);
            bson_value_destroy (&new_key);
```

Initializing a `bson_value_t` with a binary UUID requires a call to `mongoc_client_encryption_create_datakey`. It is documented in http://mongoc.org/libmongoc/current/mongoc_client_encryption_create_datakey.html

> keyid is always initialized (even on error). Caller must call [bson_value_destroy()](http://mongoc.org/libbson/current/bson_value_destroy.html) on keyid to free.